### PR TITLE
Feature/compatible with mvc 3

### PR DESCRIPTION
**Pull request by RalfEggert**
Make Zend\Navigation compatible with Zend\Mvc and Zend\Router

### CODE REVIEW

File: src/Page/Mvc.php [hunk 1]
```diff
@@ -10,8 +10,8 @@
 namespace Zend\Navigation\Page;
 
 use Zend\Mvc\ModuleRouteListener;
-use Zend\Mvc\Router\RouteMatch;
-use Zend\Mvc\Router\RouteStackInterface;
+use Zend\Router\Http\RouteMatch;
+use Zend\Router\Http\TreeRouteStack;
```
Comment (froschdesign) [inline src/Page/Mvc.php]: Can you explain, why you use a concrete implementation and no longer the interface?

Comment (RalfEggert) [inline src/Page/Mvc.php]: Sorry, my mistake. Fixed that. I blame it on the weather.... ;-)

File: src/Page/Mvc.php [hunk 2]
```diff
@@ -10,8 +10,8 @@
 namespace Zend\Navigation\Page;
 
 use Zend\Mvc\ModuleRouteListener;
-use Zend\Mvc\Router\RouteMatch;
-use Zend\Mvc\Router\RouteStackInterface;
+use Zend\Router\Http\RouteMatch;
```
Comment (froschdesign) [inline src/Page/Mvc.php]: Please change also `Zend\Router\Http\RouteMatch` to `Zend\Router\RouteMatch`.

Comment (RalfEggert) [inline src/Page/Mvc.php]: Done

**Comment by froschdesign** [comment]
@RalfEggert 
Don't forget the unit tests! (See Travis)

**Comment by RalfEggert** [comment]
Fixed the tests

**Comment by froschdesign** [comment]
@RalfEggert 
Some more:

- We have to check the the dependency to `Zend\Mvc`. Maybe we can remove it and add only `Zend\Router` to `composer.json`.
- I think, at the moment we must move `Zend\Permissions` from the "suggest" part to "require" in `composer.json`. (See [`Zend\Navigation\Page\AbstractPage`](https://github.com/zendframework/zend-navigation/blob/master/src/Page/AbstractPage.php#L15))

**Comment by vaclavvanik** [comment]
@RalfEggert you can look at https://github.com/zendframework/zend-view/blob/master/src/Helper/Url.php

It consumes both Zend\Mvc\Router and Zend\Router.

**Comment by froschdesign** [comment]
@vaclavvanik 
Please move this problem to the `Zend\View` issue tracker. Thanks!

**Comment by vaclavvanik** [comment]
@froschdesign I meant @RalfEggert can look at Zend\View\Helper\Url because this helper could be used with both Zend\Mvc 2.0 and 3.0. Ralfs changes are Mvc 3.0 imho.

**Comment by froschdesign** [comment]
@vaclavvanik 
You are right. I thought you mean the `Zend\Navigation` view helpers. :grin:

> It consumes both Zend\Mvc\Router and Zend\Router.

This is an option for a next minor release in version 2 of `Zend\Navigation`. For version 3 with can remove `Zend\Mvc` – after a deeper look into the code base.

**Comment by RalfEggert** [comment]
The only dependency to Zend\Mvc left is in https://github.com/zendframework/zend-navigation/blob/develop/src/Page/Mvc.php  It uses the two constants `ModuleRouteListener::ORIGINAL_CONTROLLER` and `ModuleRouteListener::MODULE_NAMESPACE`. 

Not sure about the `Zend\Permissions`. It is used in https://github.com/zendframework/zend-navigation/blob/develop/src/Page/AbstractPage.php and already added to the `require-dev` section in composer.json.

**Comment by weierophinney** [comment]
@RalfEggert I noticed that Travis was reporting test failures for zend-mvc v2 (and servicemanager, etc.). I took the liberty of pulling locally and figuring out how to get things running. This included:

- adding zend-router to the development requirements
- keeping zend-mvc as a development requirement, but upping it to allow either v2 or v3.
- updating the `Mvc` page type to inline the zend-mvc `ModuleRouteListener` constants, which removes the concrete dependency on zend-mvc. It's still needed for tests, as we have some integration tests with that class.
- updating the suggestions to remove zend-mvc and add zend-router, and make them more descriptive of when and why you'd need them, as well as what versions are supported.
- updating the `Mvc` page type tests to vary what type of router artifacts to use based on what version of zend-mvc is installed; if zend-mvc v2 is installed, it uses `Zend\Mvc\Router\*` classes, while for v3 it uses `Zend\Router\*` classes.

I've tested locally with both an updated `composer.lock`, as well as running `--prefer-lowest`, and it appears to work well! I'm opening a competing PR that incorporates your work just to ensure that tests pass against travis; if they do, I'll issue a 2.8.0 release.

Thanks for doing the heavy-lifting, @RalfEggert !

**Comment by RalfEggert** [comment]
:+1:

**Pull request merged**